### PR TITLE
feat: add support for dev dep groups

### PR DESCRIPTION
**Pull request by jk05**
- [x] Tests written and linted
- [ ] Documentation written / README.md updated [https://snyk.io/docs/snyk-for-node/](i)
- [x] Follows [CONTRIBUTING agreement](CONTRIBUTING.md)
- [x] Commit history is tidy [https://git-scm.com/book/en/v2/Git-Branching-Rebasing](i)
- [ ] Reviewed by Snyk team

### What this does

Dependency group notation for dev-deps is preferred since Poetry 1.2.0 (see https://python-poetry.org/docs/master/managing-dependencies). This work adds functionality to support this group notation and maintain backwards compatibility with the previous 'dev-dependencies' notation. In this iteration and in line with Poetry docs, we will handle all non-(implicit)main group deps as dev-deps.

### Notes for the reviewer

- not sure if all the test fixtures fit what is needed
- The utils for getting dev dep names may be missing things so please comment
- Had to pin `@types/babel__traverse` as I was hitting [this issue](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/63431) when trying to run locally (haven't looked into it too much at the moment)

### More information

- [OSM-12](https://snyksec.atlassian.net/browse/OSM-12)

[OSM-12]: https://snyksec.atlassian.net/browse/OSM-12?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -38,7 +38,8 @@
   "devDependencies": {
     "@types/debug": "^4.1.5",
     "@types/jest": "^25.2.3",
-    "@types/node": "^10",
+    "@types/node": "^12",
+    "@types/babel__traverse": "7.0.4",
```
Comment (jk05) [inline package.json:42]: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/63431 - had to pin it to get it working locally. Haven't looked into it too much at the moment

File: lib/manifest-parser.ts [hunk 2]
```diff
@@ -36,11 +36,37 @@ export function getDependencyNamesFrom(
...
+  for (const key in obj) {
+    if (!obj[key] || typeof obj[key] !== 'object') {
+      arr.push(key);
+      break;
+    }
+    getGroupDevDepNames(obj[key] as GroupProperty, arr);
+  }
+
+  return arr;
+};
```
Comment (jk05) [inline lib/manifest-parser.ts:52]: I was presuming having to recursively check very nested objects (I initially thought you could have `manifest.tool.poetry.group.some.nested.group.name` but seems you can only have `manifest.tool.poetry.group.name`.

File: package.json (hunk 1 above)
Comment (jk05) [reply package.json:42]: worth checking if you can get this running by removing this

**Comment by JamesPatrickGill** [comment]
I think is good to merge but should look at two things: 
* Adding the scope to the depgraph nodes a la [nodejs-lockfile-parser](https://github.com/snyk/nodejs-lockfile-parser/blob/f2e344da636ba161a8c7dbe7c8b6c113a234f82d/lib/dep-graph-builders/util.ts#L31-L32)
* Maybe use the --include-dev-deps flag/option to gate in future

File: package.json (hunk 1 above)
Comment (jk05) [reply package.json:42]: Can open another PR to bump packages and ultimately remove this

**Comment by jk05** [comment]
> I think is good to merge but should look at two things:
> 
> * Adding the scope to the depgraph nodes a la [nodejs-lockfile-parser](https://github.com/snyk/nodejs-lockfile-parser/blob/f2e344da636ba161a8c7dbe7c8b6c113a234f82d/lib/dep-graph-builders/util.ts#L31-L32)

https://github.com/snyk/snyk-poetry-lockfile-parser/pull/17  - draft currently, needs eyes

**Pull request merged**